### PR TITLE
Revoke Access grant funding permissions on permanent delivery failures

### PR DESCRIPTION
**Pull request by samuelhwilliams**
## 🎫 Ticket
https://mhclgdigital.atlassian.net/browse/FSPT-1356

## 📝 Description
When GOV.UK Notify tells us that an email address is undeliverable, we manually revoke that user's permissions.

Now that we've integrated callbacks from Notify we can take this action automatically to reduce manual intervention/chores.

We will revoke all permissions for Access grant funding users and suppress the error, unless by revoking those permissions we leave any grant without another person who has roles covering theirs. If they were the last data provider or certifier, we will log an error to flag manual intervention required still.

## 📸 Show the thing (screenshots, gifs)
```
funding-service-1  | 2026-05-12 12:29:20,367 INFO - --- POST https://funding.communities.gov.localhost:8080/deliver/api/v1/govuk-notify-callback - from before_request() in logging.py:118
funding-service-1  | 2026-05-12 12:29:20,375 INFO - audit_event: AuditEventType.SYSTEM by funding-service-notify@communities.gov.uk - from track_audit_event() in audit.py:161
funding-service-1  | 2026-05-12 12:29:20,378 ERROR - Grant 'Cheeseboards in parks' (caf0ce3f-5175-f69c-66a9-41e2c2245845) at organisation 'Ministry of Testing' (74b2ee5d-20e6-4ab9-b5d0-150444bf9510) has no other users with CERTIFIER after permanent failure; manual intervention required. - from _log_error_for_access_grant_funding_roles_with_no_alternate_users() in callbacks.py:138
funding-service-1  | 2026-05-12 12:29:20,379 INFO - 202 POST https://funding.communities.gov.localhost:8080/deliver/api/v1/govuk-notify-callback - [real:0.01s] [process:0.01s] - from after_request() in logging.py:141
funding-service-1  | 2026-05-12 12:29:30,798 INFO - --- POST https://funding.communities.gov.localhost:8080/deliver/api/v1/govuk-notify-callback - from before_request() in logging.py:118
funding-service-1  | 2026-05-12 12:29:30,808 INFO - audit_event: AuditEventType.SYSTEM by funding-service-notify@communities.gov.uk - from track_audit_event() in audit.py:161
funding-service-1  | 2026-05-12 12:29:30,808 INFO - audit_event: AuditEventType.SYSTEM by funding-service-notify@communities.gov.uk - from track_audit_event() in audit.py:161
funding-service-1  | 2026-05-12 12:29:30,808 INFO - audit_event: AuditEventType.SYSTEM by funding-service-notify@communities.gov.uk - from track_audit_event() in audit.py:161
funding-service-1  | 2026-05-12 12:29:30,810 ERROR - Grant 'Cheeseboards in parks' (caf0ce3f-5175-f69c-66a9-41e2c2245845) at organisation 'Ministry of Testing' (74b2ee5d-20e6-4ab9-b5d0-150444bf9510) has no other users with DATA_PROVIDER after permanent failure; manual intervention required. - from _log_error_for_access_grant_funding_roles_with_no_alternate_users() in callbacks.py:138
funding-service-1  | 2026-05-12 12:29:30,815 INFO - 202 POST https://funding.communities.gov.localhost:8080/deliver/api/v1/govuk-notify-callback - [real:0.02s] [process:0.01s] - from after_request() in logging.py:141
```

image

image

## 🧪 Testing
- Manually create some user roles for seeded users giving them Access grant funding permissions (eg for darin.lewis@test.communities.gov.uk with data provider/certifier for Ministry of Testing+Cheeseboards in Parks).
- Comment out this line:
    ```python
            if callback_data.to.endswith(current_app.config["GOVUK_NOTIFY_IGNORE_CALLBACK_DOMAINS"]):
                return jsonify(), 202
    ```
- POST to the /deliver/api/v1/govuk-notify-callback endpoint with `Authorisation: Bearer local-use-secret` header and payload:
    ```
    {
    "completed_at": "2026-05-12T07:46:45.250646Z",
    "created_at": "2026-05-12T07:46:44.422635Z",
    "id": "1fdbe5d2-f7e4-4a8e-af8d-592c943cfd0f",
    "notification_type": "email",
    "reference": null,
    "sent_at": "2026-05-12T07:46:44.564077Z",
    "status": "permanent-failure",
    "template_id": "e511c0d0-2ac8-4ded-80a2-13b79023c5d5",
    "template_version": 13,
    "to": "darin.lewis@test.communities.gov.uk"
    }
    ```
- Check that all user roles for that user were dropped.
- Check that error log messages were emitted if they were the last DATA_PROVIDER/CERTIFIER.
- Check that log messages are not emitted if other users have permissions covering the same org+grant+role.
- Check that users with any Deliver grant funding permissions do not have any permissions revoked.

### Review Readiness
- [x] PR title and description provide sufficient context for reviewers
- [x] Commits are logical, self-contained, and have clear descriptions

### Testing
- [x] I have tested this change and it meets the acceptance criteria for the ticket
- [x] I need the reviewer(s) to pull and run this change locally
- [x] New (non-developer) functionality has appropriate unit and integration tests
- [ ] End-to-end tests have been updated (if applicable)
- [ ] Edge cases and error conditions are tested

### CODE REVIEW

File: tests/integration/deliver_grant_funding/routes/api/test_callbacks.py [hunk 1]
```diff
@@ -133,3 +149,247 @@ def test_non_json_returns_400(self, anonymous_client) -> None:
...
+
+    class TestPermanentFailureCallback:
+        @staticmethod
+        def _post(anonymous_client, *, status: str, to: str, notification_id: uuid.UUID | None = None) -> tuple:
+            payload = {
+                "id": str(notification_id or uuid.uuid4()),
+                "reference": None,
+                "notification_type": "email",
+                "template_id": str(uuid.uuid4()),
+                "template_version": 1,
```
[the remark this answers is not in the document]
Comment (samuelhwilliams) [reply tests/integration/deliver_grant_funding/routes/api/test_callbacks.py:161]: not true

File: app/deliver_grant_funding/routes/api/callbacks.py [hunk 2]
```diff
@@ -36,7 +51,136 @@ class GovukNotifyCallbackModel(BaseModel):
...
+        current_app.logger.error(
+            "GOV.UK Notify permanent failure for unknown user: %(recipient_email)s",
+            dict(recipient_email=recipient_email),
+        )
+        return
+
+    if AuthorisationHelper.is_deliver_grant_funding_user(user):
+        # Specifically ignore Deliver grant funding users for now - TBD what we should do in the future.
+        current_app.logger.info(
+            "GOV.UK Notify permanent failure for Deliver grant funding user: %(user_id)s. No action taken.",
```
Comment (srh-sloan) [inline app/deliver_grant_funding/routes/api/callbacks.py:66]: Should this log email not id? Easier to find out who it relates to then. Or did we agree to not even be alerted to/do anything at all with deliver users that we can't email?

File: tests/integration/deliver_grant_funding/routes/api/test_callbacks.py [hunk 3]
```diff
@@ -133,3 +149,247 @@ def test_non_json_returns_400(self, anonymous_client) -> None:
...
+                permissions=[RoleEnum.DATA_PROVIDER, RoleEnum.CERTIFIER],
+            )
+
+            with caplog.at_level(logging.ERROR, logger="app"):
+                response, payload = self._post(anonymous_client, status="permanent-failure", to="bouncer@example.com")
+
+            assert response.status_code == 202
+
+            remaining = db.session.scalars(select(UserRole).where(UserRole.user_id == affected_user.id)).all()
+            assert remaining == []
```
Comment (srh-sloan) [inline tests/integration/deliver_grant_funding/routes/api/test_callbacks.py:203]: nit: Can we check that the roles for `other_user` still exist?

File: tests/integration/deliver_grant_funding/routes/api/test_callbacks.py [hunk 4]
```diff
@@ -133,3 +149,247 @@ def test_non_json_returns_400(self, anonymous_client) -> None:
...
+            )
+            factories.user_role.create(
+                user=backup_user, organisation=recipient_org, grant=None, permissions=[RoleEnum.CERTIFIER]
+            )
+
+            with caplog.at_level(logging.ERROR, logger="app"):
+                response, _ = self._post(anonymous_client, status="permanent-failure", to="bouncer@example.com")
+
+            assert response.status_code == 202
+            assert _manual_intervention_logs(caplog) == []
```
Comment (srh-sloan) [inline tests/integration/deliver_grant_funding/routes/api/test_callbacks.py:263]: nit: Should we still check the roles are all in the right state though?

**Review by srh-sloan — COMMENTED** [review]
Couple of little nits but it's all working as expected locally 👍

File: app/deliver_grant_funding/routes/api/callbacks.py (hunk 2 above)
Comment (samuelhwilliams) [reply app/deliver_grant_funding/routes/api/callbacks.py:66]: One of our principles somewhere is to just log user IDs rather than direct personal data, so trying to stay in line with that 🙏

**Review by srh-sloan — APPROVED** [review]
All works locally. Really nice testing instructions thank you 🙏
